File: backend/services/mobile_navigation.py
"""账号级手机导航配置的校验与兼容处理。"""

from __future__ import annotations

import json
from typing import Any, Iterable


DEFAULT_MOBILE_NAVIGATION_MODE = "dock"
MAX_DOCK_GROUPS = 4
DOCK_CONFIG_VERSION = 2

GROUP_ITEMS: dict[str, tuple[str, ...]] = {
    "workspace": (
        "dashboard",
        "online_query",
        "data_upload",
        "work_log",
        "workflow_tickets",
    ),
    "tasks": (
        "flow_tasks",
        "police_tasks",
        "police_analysis",
        "photo_tasks",
    ),
    "summaries": (
        "online_summary",
        "visit_summary",
    ),
    "resources": (
        "grid_members",
        "communities",
        "police_addresses",
        "registry",
        "watch_people",
        "users",
        "permission_groups",
    ),
    "system": (
        "settings",
        "workflow_config",
        "operations",
    ),
}

SUPER_ADMIN_ITEMS = {"users", "permission_groups", "operations"}
ADMIN_ITEMS = {"data_upload", "work_log", "police_addresses"}

ITEM_PERMISSIONS: dict[str, str] = {
    "online_summary": "online.summary.view",
    "online_query": "online.raw.view",
    "flow_tasks": "online.raw.view",
    "police_tasks": "police.dispatch.manage",
    "police_analysis": "online.task.manage",
    "photo_tasks": "workflow.ticket.handle",
    "workflow_tickets": "workflow.ticket.view",
    "visit_summary": "visit.summary.view",
    "data_upload": "visit.import",
    "work_log": "worklog.manage",
    "grid_members": "personnel.basic.view",
    "communities": "community.view",
    "police_addresses": "police.address.manage",
    "registry": "registry.property.view",
    "watch_people": "registry.watch.view",
    "users": "user.manage",
    "permission_groups": "permission.manage",
    "workflow_config": "workflow.config.manage",
    "operations": "ops.manage",
}

ITEM_PERMISSION_ALTERNATIVES: dict[str, tuple[str, ...]] = {
    "data_upload": ("visit.import", "police.dispatch.manage"),
    "police_analysis": ("online.task.manage", "police.dispatch.manage"),
    "photo_tasks": ("workflow.ticket.handle", "workflow.ticket.manage"),
}


def normalize_mobile_navigation_mode(value: Any) -> str:
    return (
        str(value)
        if value in {"sidebar", "dock"}
        else DEFAULT_MOBILE_NAVIGATION_MODE
    )


def _admin_code_access(
    role: str,
    permission_group_codes: Iterable[str] | None,
) -> bool:
    return role in {"admin", "super_admin"} or bool(
        {str(code).strip() for code in permission_group_codes or []}
        & {"admin", "super_admin"}
    )


def _item_is_accessible(
    item_id: str,
    role: str,
    permissions: Iterable[str] | None = None,
    permission_group_codes: Iterable[str] | None = None,
    position: str | None = None,
) -> bool:
    admin_access = _admin_code_access(role, permission_group_codes)
    if item_id == "online_query":
        group_codes = {
            str(code).strip() for code in permission_group_codes or []
        }
        query_admin_access = bool(group_codes & {"admin", "super_admin"}) or (
            not group_codes and role in {"admin", "super_admin"}
        )
        if not query_admin_access:
            return False
    if item_id == "flow_tasks":
        has_task_manage = permissions is not None and "online.task.manage" in permissions
        if position not in {"组长", "组员"} and not has_task_manage and not admin_access:
            return False
    if item_id == "police_tasks":
        if position not in {"基础管控", "中队长"} and not (not position and admin_access):
            return False
    if item_id == "police_analysis":
        has_analysis_permission = permissions is not None and any(
            permission in permissions
            for permission in ITEM_PERMISSION_ALTERNATIVES["police_analysis"]
        )
        if not has_analysis_permission and not admin_access:
            return False
    if item_id == "photo_tasks":
        has_manage_permission = permissions is not None and "workflow.ticket.manage" in permissions
        if position != "基础管控" and not has_manage_permission and not admin_access:
            return False
    if permissions is not None:
        alternatives = ITEM_PERMISSION_ALTERNATIVES.get(item_id)
        if alternatives:
            return any(permission in permissions for permission in alternatives)
        required = ITEM_PERMISSIONS.get(item_id)
        return required is None or required in permissions
    if item_id in SUPER_ADMIN_ITEMS:
        return role == "super_admin"
    if item_id in ADMIN_ITEMS:
        return role in {"admin", "super_admin"}
    return True


def default_mobile_dock_config(
    role: str,
    permissions: Iterable[str] | None = None,
    permission_group_codes: Iterable[str] | None = None,
    position: str | None = None,
) -> dict[str, Any]:
    """返回新版本默认 Dock；所有岗位按同一分类顺序取前四组。"""
    groups = []
    for group_id, item_ids in GROUP_ITEMS.items():
        items = [
            item_id
            for item_id in item_ids
            if _item_is_accessible(
                item_id,
                role,
                permissions,
                permission_group_codes,
                position,
            )
        ]
        if items:
            groups.append({"id": group_id, "items": items})
    return {"version": DOCK_CONFIG_VERSION, "groups": groups[:MAX_DOCK_GROUPS]}


def _load_config(value: Any) -> dict[str, Any] | None:
    if value is None:
        return None
    if isinstance(value, bytes):
        value = value.decode("utf-8", errors="replace")
    if isinstance(value, str):
        if not value.strip():
            return None
        try:
            value = json.loads(value)
        except (TypeError, ValueError):
            return None
    return value if isinstance(value, dict) else None


def normalize_mobile_dock_config(
    value: Any,
    role: str,
    permissions: Iterable[str] | None = None,
    permission_group_codes: Iterable[str] | None = None,
    position: str | None = None,
) -> dict[str, Any]:
    """过滤配置；旧版本配置故意不迁移，直接返回 v2 默认值。"""
    parsed = _load_config(value)
    if not parsed or parsed.get("version") != DOCK_CONFIG_VERSION:
        return default_mobile_dock_config(
            role, permissions, permission_group_codes, position
        )
    raw_groups = parsed.get("groups")
    if not isinstance(raw_groups, list):
        return default_mobile_dock_config(
            role, permissions, permission_group_codes, position
        )

    raw_task_items = {
        str(item or "")
        for group in raw_groups
        if isinstance(group, dict)
        and str(group.get("id") or "") == "tasks"
        and isinstance(group.get("items"), list)
        for item in group["items"]
    }
    groups: list[dict[str, Any]] = []
    seen_groups: set[str] = set()
    for raw_group in raw_groups[:MAX_DOCK_GROUPS]:
        if not isinstance(raw_group, dict):
            continue
        group_id = str(raw_group.get("id") or "")
        if group_id not in GROUP_ITEMS or group_id in seen_groups:
            continue
        raw_items = raw_group.get("items")
        if not isinstance(raw_items, list):
            continue
        allowed = set(GROUP_ITEMS[group_id])
        seen_items: set[str] = set()
        items: list[str] = []
        for raw_item in raw_items:
            item_id = str(raw_item or "")
            if (
                item_id not in allowed
                or item_id in seen_items
                or not _item_is_accessible(
                    item_id,
                    role,
                    permissions,
                    permission_group_codes,
                    position,
                )
            ):
                continue
            seen_items.add(item_id)
            items.append(item_id)
        if items:
            seen_groups.add(group_id)
            groups.append({"id": group_id, "items": items})

    accessible_defaults = default_mobile_dock_config(
        role, permissions, permission_group_codes, position
    )["groups"]
    if not groups:
        return {
            "version": DOCK_CONFIG_VERSION,
            "groups": accessible_defaults,
        }
    workspace = next((group for group in groups if group["id"] == "workspace"), None)
    if workspace is None:
        workspace = {"id": "workspace", "items": ["dashboard"]}
        groups.insert(0, workspace)
    else:
        workspace["items"] = [
            "dashboard",
            *[item for item in workspace["items"] if item != "dashboard"],
        ]
        groups = [workspace, *[group for group in groups if group is not workspace]]
    accessible_by_group = {
        group["id"]: set(group["items"])
        for group in accessible_defaults
    }
    if (
        "workflow_tickets" in raw_task_items
        and "workflow_tickets" in accessible_by_group.get("workspace", set())
        and "workflow_tickets" not in workspace["items"]
    ):
        workspace["items"].append("workflow_tickets")
    tasks = next((group for group in groups if group["id"] == "tasks"), None)
    if tasks is not None:
        for previous_id, next_id in (
            ("police_tasks", "police_analysis"),
            ("workflow_tickets", "photo_tasks"),
        ):
            if (
                previous_id in raw_task_items
                and next_id in accessible_by_group.get("tasks", set())
                and next_id not in tasks["items"]
            ):
                tasks["items"].append(next_id)
    present = {group["id"] for group in groups}
    for candidate in accessible_defaults:
        if len(groups) >= MAX_DOCK_GROUPS:
            break
        if candidate["id"] in present:
            continue
        groups.append(candidate)
        present.add(candidate["id"])
    return {"version": DOCK_CONFIG_VERSION, "groups": groups[:MAX_DOCK_GROUPS]}


def validate_mobile_dock_config(
    value: Any,
    role: str,
    permissions: Iterable[str] | None = None,
    permission_group_codes: Iterable[str] | None = None,
    position: str | None = None,
) -> dict[str, Any]:
    """保存配置前做严格校验，避免静默接受错误或越权入口。"""
    parsed = _load_config(value)
    if not parsed or parsed.get("version") != DOCK_CONFIG_VERSION:
        raise ValueError("Dock 配置版本过旧，请刷新后重试")
    raw_groups = parsed.get("groups")
    if not isinstance(raw_groups, list):
        raise ValueError("Dock 配置格式不正确")
    if not 1 <= len(raw_groups) <= MAX_DOCK_GROUPS:
        raise ValueError("Dock 必须保留 1 至 4 个分类")

    groups: list[dict[str, Any]] = []
    seen_groups: set[str] = set()
    for raw_group in raw_groups:
        if not isinstance(raw_group, dict):
            raise ValueError("Dock 分类格式不正确")
        group_id = str(raw_group.get("id") or "")
        if group_id not in GROUP_ITEMS:
            raise ValueError(f"未知的 Dock 分类：{group_id or '空值'}")
        if group_id in seen_groups:
            raise ValueError("Dock 分类不能重复")
        raw_items = raw_group.get("items")
        if not isinstance(raw_items, list) or not raw_items:
            raise ValueError("每个 Dock 分类至少保留一个页面")
        allowed = set(GROUP_ITEMS[group_id])
        seen_items: set[str] = set()
        items: list[str] = []
        for raw_item in raw_items:
            item_id = str(raw_item or "")
            if item_id not in allowed:
                raise ValueError(f"页面不属于当前分类：{item_id or '空值'}")
            if item_id in seen_items:
                raise ValueError("同一分类中的页面不能重复")
            if not _item_is_accessible(
                item_id,
                role,
                permissions,
                permission_group_codes,
                position,
            ):
                raise ValueError("Dock 配置包含当前账号无权访问的页面")
            seen_items.add(item_id)
            items.append(item_id)
        seen_groups.add(group_id)
        groups.append({"id": group_id, "items": items})

    workspace = next((group for group in groups if group["id"] == "workspace"), None)
    if workspace is None:
        groups.insert(0, {"id": "workspace", "items": ["dashboard"]})
    else:
        workspace["items"] = [
            "dashboard",
            *[item for item in workspace["items"] if item != "dashboard"],
        ]
        groups = [workspace, *[group for group in groups if group is not workspace]]
    present = {group["id"] for group in groups}
    for candidate in default_mobile_dock_config(
        role, permissions, permission_group_codes, position
    )["groups"]:
        if len(groups) >= MAX_DOCK_GROUPS:
            break
        if candidate["id"] in present:
            continue
        groups.append(candidate)
        present.add(candidate["id"])
    return {"version": DOCK_CONFIG_VERSION, "groups": groups[:MAX_DOCK_GROUPS]}


def serialize_mobile_dock_config(value: dict[str, Any]) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))

File: frontend/src/navigation/mobileNavigation.ts
import type {
  MobileDockConfig,
  MobileNavigationGroupId,
  MobileNavigationItemId,
  PermissionCode,
  Role,
} from '../types'

export type NavigationIconName =
  | 'dashboard'
  | 'workspace'
  | 'resources'
  | 'system'
  | 'summary'
  | 'query'
  | 'visit'
  | 'upload'
  | 'worklog'
  | 'members'
  | 'communities'
  | 'users'
  | 'settings'
  | 'operations'

export interface NavigationItemDefinition {
  id: MobileNavigationItemId
  path: string
  label: string
  shortLabel: string
  icon: NavigationIconName
  end?: boolean
  roles?: Role[]
  permission?: PermissionCode
  anyPermissions?: PermissionCode[]
}

export interface NavigationGroupDefinition {
  id: MobileNavigationGroupId
  label: string
  dockLabel: string
  icon: NavigationIconName
  items: NavigationItemDefinition[]
}

export const NAVIGATION_GROUPS: NavigationGroupDefinition[] = [
  {
    id: 'workspace',
    label: '工作台',
    dockLabel: '工作台',
    icon: 'workspace',
    items: [
      {
        id: 'dashboard',
        path: '/',
        label: '仪表盘',
        shortLabel: '首页',
        icon: 'dashboard',
        end: true,
      },
      {
        id: 'online_query',
        path: '/query',
        label: '在线数据查询',
        shortLabel: '在线查询',
        icon: 'query',
        permission: 'online.raw.view',
      },
      {
        id: 'data_upload',
        path: '/data-upload',
        label: '数据上传中心',
        shortLabel: '数据上传',
        icon: 'upload',
        anyPermissions: ['visit.import', 'police.dispatch.manage'],
        roles: ['super_admin', 'admin'],
      },
      {
        id: 'work_log',
        path: '/work-log',
        label: '文件生成',
        shortLabel: '文件生成',
        icon: 'worklog',
        permission: 'worklog.manage',
        roles: ['super_admin', 'admin'],
      },
      {
        id: 'workflow_tickets',
        path: '/workflow',
        label: '工单中心',
        shortLabel: '工单',
        icon: 'worklog',
        permission: 'workflow.ticket.view',
      },
    ],
  },
  {
    id: 'tasks',
    label: '任务处理',
    dockLabel: '任务',
    icon: 'query',
    items: [
      {
        id: 'flow_tasks',
        path: '/tasks/home',
        label: '流口指令核查',
        shortLabel: '流口核查',
        icon: 'query',
        permission: 'online.raw.view',
      },
      {
        id: 'police_tasks',
        path: '/police-tasks',
        label: '下发任务处理',
        shortLabel: '下发任务',
        icon: 'worklog',
        permission: 'police.dispatch.manage',
      },
      {
        id: 'police_analysis',
        path: '/police-analysis',
        label: '研判',
        shortLabel: '研判',
        icon: 'worklog',
        anyPermissions: ['online.task.manage', 'police.dispatch.manage'],
      },
      {
        id: 'photo_tasks',
        path: '/photo-tasks',
        label: '调照片',
        shortLabel: '调照片',
        icon: 'worklog',
        anyPermissions: ['workflow.ticket.handle', 'workflow.ticket.manage'],
      },
    ],
  },
  {
    id: 'summaries',
    label: '统计汇总',
    dockLabel: '汇总',
    icon: 'summary',
    items: [
      {
        id: 'online_summary',
        path: '/summary',
        label: '在线数据汇总',
        shortLabel: '在线汇总',
        icon: 'summary',
        end: true,
        permission: 'online.summary.view',
      },
      {
        id: 'visit_summary',
        path: '/visit-summary',
        label: '走访汇总',
        shortLabel: '走访汇总',
        icon: 'visit',
        permission: 'visit.summary.view',
      },
    ],
  },
  {
    id: 'resources',
    label: '基础资料',
    dockLabel: '基础资料',
    icon: 'resources',
    items: [
      {
        id: 'grid_members',
        path: '/grid-members',
        label: '人员管理',
        shortLabel: '人员管理',
        icon: 'members',
        permission: 'personnel.basic.view',
      },
      {
        id: 'communities',
        path: '/communities',
        label: '社区管理',
        shortLabel: '社区管理',
        icon: 'communities',
        permission: 'community.view',
      },
      {
        id: 'police_addresses',
        path: '/police-addresses',
        label: '小区管理',
        shortLabel: '小区管理',
        icon: 'communities',
        permission: 'police.address.manage',
      },
      {
        id: 'registry',
        path: '/registry',
        label: '辖区档案',
        shortLabel: '辖区档案',
        icon: 'resources',
        permission: 'registry.property.view',
      },
      {
        id: 'watch_people',
        path: '/watch-people',
        label: '人员标记',
        shortLabel: '人员标记',
        icon: 'members',
        permission: 'registry.watch.view',
      },
      {
        id: 'users',
        path: '/users',
        label: '用户管理',
        shortLabel: '用户管理',
        icon: 'users',
        permission: 'user.manage',
        roles: ['super_admin'],
      },
      {
        id: 'permission_groups',
        path: '/permission-groups',
        label: '权限组管理',
        shortLabel: '权限组',
        icon: 'users',
        permission: 'permission.manage',
        roles: ['super_admin'],
      },
    ],
  },
  {
    id: 'system',
    label: '设置',
    dockLabel: '设置',
    icon: 'system',
    items: [
      {
        id: 'settings',
        path: '/settings',
        label: '设置',
        shortLabel: '系统设置',
        icon: 'settings',
      },
      {
        id: 'workflow_config',
        path: '/settings/workflow',
        label: '工单流程配置',
        shortLabel: '流程配置',
        icon: 'worklog',
        permission: 'workflow.config.manage',
        roles: ['super_admin'],
      },
      {
        id: 'operations',
        path: '/operations',
        label: '运维中心',
        shortLabel: '运维中心',
        icon: 'operations',
        permission: 'ops.manage',
        roles: ['super_admin'],
      },
    ],
  },
]

export const MAX_DOCK_GROUPS = 4

export function isNavigationItemAccessible(
  item: NavigationItemDefinition,
  role: Role,
  permissions?: PermissionCode[],
  permissionGroupCodes: string[] = [],
  position?: string | null,
): boolean {
  const adminAccess = ['admin', 'super_admin'].includes(role)
    || permissionGroupCodes.some(code => ['admin', 'super_admin'].includes(code))
  const queryAdminAccess = permissionGroupCodes.length > 0
    ? permissionGroupCodes.some(code => ['admin', 'super_admin'].includes(code))
    : ['admin', 'super_admin'].includes(role)
  if (item.id === 'online_query' && !queryAdminAccess) return false
  if (
    item.id === 'flow_tasks'
    && !['组长', '组员', '片长', '基础管控', '中队长', '社区民警', '所队领导'].includes(position || '')
    && !permissions?.includes('online.task.manage')
    && !adminAccess
  ) return false
  if (
    item.id === 'police_tasks'
    && !['基础管控', '中队长'].includes(position || '')
    && !(
      !position
      && (
        adminAccess
      )
    )
  ) return false
  if (
    item.id === 'police_analysis'
    && !permissions?.some(permission => (
      permission === 'online.task.manage' || permission === 'police.dispatch.manage'
    ))
    && !adminAccess
  ) return false
  if (
    item.id === 'photo_tasks'
    && position !== '基础管控'
    && !permissions?.includes('workflow.ticket.manage')
    && !adminAccess
  ) return false
  if (item.permission) {
    // Permission data is authoritative.  If it is unavailable, fail closed
    // for permission-gated items instead of exposing a menu by accident.
    return permissions
      ? permissions.includes(item.permission)
      : role === 'super_admin' || Boolean(item.roles?.includes(role))
  }
  if (item.anyPermissions) {
    return permissions
      ? item.anyPermissions.some(permission => permissions.includes(permission))
      : role === 'super_admin' || Boolean(item.roles?.includes(role))
  }
  return !item.roles || item.roles.includes(role)
}

export function accessibleNavigationGroups(
  role: Role,
  permissions?: PermissionCode[],
  permissionGroupCodes: string[] = [],
  position?: string | null,
): NavigationGroupDefinition[] {
  return NAVIGATION_GROUPS.map(group => ({
    ...group,
    items: group.items.filter(item => (
      isNavigationItemAccessible(item, role, permissions, permissionGroupCodes, position)
    )),
  })).filter(group => group.items.length > 0)
}

export function defaultMobileDockConfig(
  role: Role,
  permissions?: PermissionCode[],
  permissionGroupCodes: string[] = [],
  position?: string | null,
): MobileDockConfig {
  return {
    version: 2,
    groups: accessibleNavigationGroups(role, permissions, permissionGroupCodes, position).slice(0, MAX_DOCK_GROUPS).map(group => ({
      id: group.id,
      items: group.items.map(item => item.id),
    })),
  }
}

export function normalizeMobileDockConfig(
  value: MobileDockConfig | null | undefined,
  role: Role,
  permissions?: PermissionCode[],
  permissionGroupCodes: string[] = [],
  position?: string | null,
): MobileDockConfig {
  if (!value || value.version !== 2 || !Array.isArray(value.groups)) {
    return defaultMobileDockConfig(role, permissions, permissionGroupCodes, position)
  }

  const definitions = new Map(
    accessibleNavigationGroups(role, permissions, permissionGroupCodes, position).map(group => [group.id, group]),
  )
  const rawTaskItems = new Set(
    value.groups
      .filter(group => group.id === 'tasks' && Array.isArray(group.items))
      .flatMap(group => group.items),
  )
  const seenGroups = new Set<MobileNavigationGroupId>()
  const groups = value.groups.slice(0, MAX_DOCK_GROUPS).flatMap((rawGroup) => {
    const definition = definitions.get(rawGroup.id)
    if (!definition || seenGroups.has(rawGroup.id)) return []
    const allowedItems = new Set(definition.items.map(item => item.id))
    const seenItems = new Set<MobileNavigationItemId>()
    const items = (Array.isArray(rawGroup.items) ? rawGroup.items : [])
      .flatMap((itemId) => {
        if (!allowedItems.has(itemId) || seenItems.has(itemId)) return []
        seenItems.add(itemId)
        return [itemId]
      })
    if (items.length === 0) return []
    seenGroups.add(rawGroup.id)
    return [{ id: rawGroup.id, items }]
  })

  const normalized = groups.length > 0
    ? groups
    : defaultMobileDockConfig(role, permissions, permissionGroupCodes, position).groups
  const workspaceDefinition = definitions.get('workspace')
  let workspace = normalized.find(group => group.id === 'workspace')
  if (workspaceDefinition) {
    const dashboardId: MobileNavigationItemId = 'dashboard'
    if (workspace) {
      workspace.items = [dashboardId, ...workspace.items.filter(item => item !== dashboardId)]
    } else {
      workspace = { id: 'workspace', items: [dashboardId] }
      normalized.unshift(workspace)
    }
    if (
      rawTaskItems.has('workflow_tickets')
      && workspaceDefinition.items.some(item => item.id === 'workflow_tickets')
      && !workspace.items.includes('workflow_tickets')
    ) workspace.items.push('workflow_tickets')
  }
  const tasks = normalized.find(group => group.id === 'tasks')
  const taskDefinition = definitions.get('tasks')
  if (tasks && taskDefinition) {
    const appendMovedTask = (
      previousId: MobileNavigationItemId,
      nextId: MobileNavigationItemId,
    ) => {
      if (
        rawTaskItems.has(previousId)
        && taskDefinition.items.some(item => item.id === nextId)
        && !tasks.items.includes(nextId)
      ) tasks.items.push(nextId)
    }
    appendMovedTask('police_tasks', 'police_analysis')
    appendMovedTask('workflow_tickets', 'photo_tasks')
  }
  const presentGroups = new Set(normalized.map(group => group.id))
  for (const definition of accessibleNavigationGroups(
    role,
    permissions,
    permissionGroupCodes,
    position,
  )) {
    if (normalized.length >= MAX_DOCK_GROUPS) break
    if (presentGroups.has(definition.id)) continue
    normalized.push({
      id: definition.id,
      items: definition.items.map(item => item.id),
    })
    presentGroups.add(definition.id)
  }
  return {
    version: 2,
    groups: [
      ...normalized.filter(group => group.id === 'workspace'),
      ...normalized.filter(group => group.id !== 'workspace'),
    ].slice(0, MAX_DOCK_GROUPS),
  }
}

function moveEntry<T>(items: T[], fromIndex: number, toIndex: number): T[] {
  if (
    fromIndex < 0
    || toIndex < 0
    || fromIndex >= items.length
    || toIndex >= items.length
    || fromIndex === toIndex
  ) return items
  const next = [...items]
  const [entry] = next.splice(fromIndex, 1)
  next.splice(toIndex, 0, entry)
  return next
}

export function reorderMobileDockGroups(
  config: MobileDockConfig,
  activeGroupId: MobileNavigationGroupId,
  overGroupId: MobileNavigationGroupId,
): MobileDockConfig {
  const fromIndex = config.groups.findIndex(group => (
    group.id === activeGroupId
  ))
  const toIndex = config.groups.findIndex(group => group.id === overGroupId)
  return {
    version: 2,
    groups: moveEntry(config.groups, fromIndex, toIndex),
  }
}

export function reorderMobileDockItems(
  config: MobileDockConfig,
  groupId: MobileNavigationGroupId,
  activeItemId: MobileNavigationItemId,
  overItemId: MobileNavigationItemId,
): MobileDockConfig {
  return {
    version: 2,
    groups: config.groups.map((group) => {
      if (group.id !== groupId) return group
      return {
        ...group,
        items: moveEntry(
          group.items,
          group.items.indexOf(activeItemId),
          group.items.indexOf(overItemId),
        ),
      }
    }),
  }
}

export function navigationGroupById(
  id: MobileNavigationGroupId,
): NavigationGroupDefinition | undefined {
  return NAVIGATION_GROUPS.find(group => group.id === id)
}

export function navigationItemById(
  id: MobileNavigationItemId,
): NavigationItemDefinition | undefined {
  return NAVIGATION_GROUPS
    .flatMap(group => group.items)
    .find(item => item.id === id)
}

export function routeIsActive(
  pathname: string,
  item: NavigationItemDefinition,
): boolean {
  if (item.id === 'flow_tasks' && pathname.startsWith('/tasks')) return true
  if (item.end) return pathname === item.path
  return pathname === item.path || pathname.startsWith(`${item.path}/`)
}

export function mobileNavigationItemLabel(
  item: NavigationItemDefinition,
  position?: string | null,
  short = false,
): string {
  if (item.id === 'dashboard') return '首页'
  void position
  return short ? item.shortLabel : item.label
}
